primitives/src/parachain.rs

```diff
@@ -164,119 +162,274 @@ pub struct DutyRoster {
+/// is necessary for validation of the parachain candidate.
 #[derive(PartialEq, Eq, Clone, Encode, Decode)]
 #[cfg_attr(feature = "std", derive(Debug, Default))]
-pub struct CandidateReceipt {
+pub struct OmittedValidationData {
+	/// The global validation schedule.
+	pub global_validation: GlobalValidationSchedule,
+	/// The local validation data.
+	pub local_validation: LocalValidationData,
+}
+
+/// An abridged candidate-receipt.
+///
+/// Much info in a candidate-receipt is duplicated from the relay-chain state.
+/// When submitting to the relay-chain, this data should be omitted as it can
+/// be re-generated from relay-chain state.
+#[derive(PartialEq, Eq, Clone, Encode, Decode)]
+#[cfg_attr(feature = "std", derive(Debug, Default))]
+pub struct AbridgedCandidateReceipt {
 	/// The ID of the parachain this is a candidate for.
 	pub parachain_index: Id,
+	/// The hash of the relay-chain block this should be executed in
+	/// the context of.
+	// NOTE: the fact that the hash includes this value means that code depends
+	// on this for deduplication. Removing this field is likely to break things.
+	pub relay_parent: Hash,
+	/// The head-data
+	pub head_data: HeadData,
 	/// The collator's relay-chain account ID
 	pub collator: CollatorId,
 	/// Signature on blake2-256 of the block data by collator.
 	pub signature: CollatorSignature,
-	/// The head-data
-	pub head_data: HeadData,
-	/// The parent head-data.
-	pub parent_head: HeadData,
-	/// Fees paid from the chain to the relay chain validators
-	pub fees: Balance,
-	/// blake2-256 Hash of block data.
-	pub block_data_hash: Hash,
-	/// Messages destined to be interpreted by the Relay chain itself.
-	pub upward_messages: Vec<UpwardMessage>,
-	/// The root of a block's erasure encoding Merkle tree.
-	pub erasure_root: Hash,
+	/// The hash of the pov-block.
+	pub pov_block_hash: Hash,
+	/// Commitments made as a result of validation.
+	pub commitments: CandidateCommitments,
 }
 
-impl CandidateReceipt {
-	/// Get the blake2_256 hash
+impl AbridgedCandidateReceipt {
+	/// Compute the hash of the abridged candidate receipt.
+	///
+	/// This is often used as the canonical hash of the receipt, rather than
+	/// the hash of the full receipt. The reason being that all data in the full
+	/// receipt is comitted to in the abridged receipt; this receipt references
+	/// the relay-chain block in which context it should be executed, which implies
+	/// any blockchain state that must be referenced.
 	pub fn hash(&self) -> Hash {
 		use runtime_primitives::traits::{BlakeTwo256, Hash};
 		BlakeTwo256::hash_of(self)
 	}
 
-	/// Check integrity vs. provided block data.
-	pub fn check_signature(&self) -> Result<(), ()> {
-		use runtime_primitives::traits::AppVerify;
+	/// Combine the abridged candidate receipt with the omitted data,
+	/// forming a full `CandidateReceipt`.
+	pub fn complete(self, omitted: OmittedValidationData) -> CandidateReceipt {
```

Comment:
Maybe `impl From<(AbridgedCandidateReceipt, OmittedValidationData)> for CandidateReceipt`? And also for other type conversions.

Reply:
The functions are almost certainly more clear & discoverable than a `From` implementation on a tuple.